### PR TITLE
MODINREACH-8 Define Data Model for INN-Reach Transactions (All Types)

### PR DESCRIPTION
**Pull request by maryna-steshenko**
## Purpose

Define Data Model for INN-Reach Transactions (All Types).
JIRA: [MODINREACH-8](https://issues.folio.org/browse/MODINREACH-8)
## Approach

- InnReachTransaction model added.
- TransactionHold models added.
- InnReachTransaction repository added.
- SQL scripts for database tables added.
- SQL scripts for tests added.
- InnReachTransaction repository tests added.
## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/innreach/domain/entity/TransactionHold.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import javax.persistence.Table;
+import java.time.OffsetDateTime;
+import java.util.UUID;
+
+@Getter
+@Setter
+@Entity
+@Inheritance(strategy = InheritanceType.JOINED)
+@Table(name = "transaction_hold")
+@ToString
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/entity/TransactionHold.java]: `pickupLocation` needs to be excluded

File: src/main/resources/db/changelog/changes/v1.0.0/sql/2021-09-07-01__create-transaction-hold-tables.sql [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    REFERENCES transaction_hold (id) ON DELETE CASCADE
+);
+
+CREATE TABLE transaction_item_hold
+(
+    id UUID NOT NULL,
+    central_patron_type SMALLINT NOT NULL CHECK(central_patron_type BETWEEN 0 AND 255),
+    patron_name VARCHAR(255) NOT NULL,
+    CONSTRAINT pk_transaction_item_hold PRIMARY KEY (id),
+    CONSTRAINT fk_item_hold_transaction FOREIGN KEY (id)
```
Comment (oliinyko) [inline src/main/resources/db/changelog/changes/v1.0.0/sql/2021-09-07-01__create-transaction-hold-tables.sql]: `fk_transaction_item_hold` to make naming more consistent

File: src/main/resources/db/changelog/changes/v1.0.0/sql/2021-09-07-02__create-inn-reach-transaction-table.sql [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    transaction_hold_id UUID NOT NULL,
+    created_by_userid UUID NOT NULL DEFAULT '00000000-0000-0000-0000-000000000000',
+    created_by_username VARCHAR(255) NOT NULL DEFAULT 'SYSTEM',
+    updated_by_userid UUID,
+    updated_by_username VARCHAR(255),
+    created_date TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    updated_date TIMESTAMP,
+    CONSTRAINT pk_inn_reach_transaction PRIMARY KEY (id),
+    CONSTRAINT fk_transaction_hold_id FOREIGN KEY (transaction_hold_id)
+    REFERENCES transaction_hold (id) ON DELETE CASCADE,
```
Comment (oliinyko) [inline src/main/resources/db/changelog/changes/v1.0.0/sql/2021-09-07-02__create-inn-reach-transaction-table.sql]: this `delete cascade` constraint probably needs to be on `transaction_hold` side

**Comment by oliinyko** [comment]
I assume that `inn_reach_transaction` is the main entity in relations, if so then it means that all the child entities needs to be removed on cascade when the transaction is removed. Please check if it works this way

**Comment by oliinyko** [comment]
According to requirements there also should be `trackingId`:

> Requirement: All records have:
 trackingId
ID for transaction tracking
*** centralCode
Unique code that identifies the central server
Provided as URL parameters to the post/put API requests that create each transaction type.

File: src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import org.folio.innreach.domain.entity.TransactionHold;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+import java.util.Optional;
+import java.util.UUID;
+
+public interface InnReachTransactionRepository extends JpaRepository<InnReachTransaction, UUID> {
+  @Query(name = InnReachTransaction.FETCH_TRANSACTION_HOLD_NAME)
+  Optional<TransactionHold> findTransactionHoldById(UUID id);
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java]: is there any use case to load transaction hold without transaction itself?

Comment (maryna-steshenko) [reply src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java]: No, there's no such use case, added this method just for test.

**Comment by maryna-steshenko** [comment]
> According to requirements there also should be `trackingId`:
> 
> > Requirement: All records have:
> > trackingId
> > ID for transaction tracking
> > *** centralCode
> > Unique code that identifies the central server
> > Provided as URL parameters to the post/put API requests that create each transaction type.

@oliinyko I suppose, entity's id can be used for transaction tracking, but maybe it's better to clarify if trackinId meant to be a separate field.

**Comment by oliinyko** [comment]
> > According to requirements there also should be `trackingId`:
> > > Requirement: All records have:
> > > trackingId
> > > ID for transaction tracking
> > > *** centralCode
> > > Unique code that identifies the central server
> > > Provided as URL parameters to the post/put API requests that create each transaction type.
> 
> @oliinyko I suppose, entity's id can be used for transaction tracking, but maybe it's better to clarify if trackinId meant to be a separate field.

Ok I see, but it's better to clarify if circulation API endpoint supports UUID string as trackingId. There potentially may be some limitations in length/type

File: src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java (hunk 4 above)
Comment (oliinyko) [reply src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java]: Let's then remove it and add a method that will load transaction with all it's related entities instead

Comment (oliinyko) [reply src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java]: Also, I saw `Hibernate.unproxy` usage in tests, let's avoid this too by using a method which loads everything needed

**Review by andrew-khodyka — APPROVED** [review]

File: src/main/resources/db/changelog/changes/v1.0.0/sql/2021-09-07-02__create-inn-reach-transaction-table.sql [hunk 5]
```diff
@@ -0,0 +1,18 @@
+CREATE TABLE inn_reach_transaction
+(
+    id UUID NOT NULL,
+    tracking_id UUID NOT NULL,
```
Comment (oliinyko) [inline src/main/resources/db/changelog/changes/v1.0.0/sql/2021-09-07-02__create-inn-reach-transaction-table.sql]: Should it be of UUID type?

Comment (maryna-steshenko) [reply src/main/resources/db/changelog/changes/v1.0.0/sql/2021-09-07-02__create-inn-reach-transaction-table.sql]: Will change it to a string type.

**Review by oliinyko — APPROVED** [review]

**Pull request merged**